tests/components/decora_wifi/test_config_flow.py

```diff
@@ -0,0 +1,216 @@
+
+
+async def test_abort_if_existing_entry(hass: HomeAssistant):
+    """Check flow abort when an entry already exist."""
+    config_entry = MockConfigEntry(
+        domain=DOMAIN,
+        unique_id=USERNAME,
+        data={CONF_USERNAME: USERNAME, CONF_PASSWORD: PASSWORD},
+    )
+    config_entry.add_to_hass(hass)
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": config_entries.SOURCE_USER},
+        data={
+            CONF_USERNAME: USERNAME,
+            CONF_PASSWORD: PASSWORD,
+        },
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+    assert result["reason"] == "already_configured"
+
+
+async def test_user_validate_invalid_login(hass: HomeAssistant):
+    """Test user flow with invalid username."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    with patch(
+        "homeassistant.components.decora_wifi.common.DecoraWiFiSession",
+        side_effect=FakeDecoraWiFiSession,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {CONF_USERNAME: USERNAME, CONF_PASSWORD: INCORRECT_PASSWORD},
+        )
+    assert result2["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result2["step_id"] == "user_validate"
+    assert result2["errors"] == {"base": "invalid_auth"}
+
+
+async def test_user_validate_no_internet_connection(hass: HomeAssistant):
+    """Test user flow with no internet connection."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.decora_wifi.common.DecoraWiFiSession.login",
+        side_effect=CommFailed(),
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {CONF_USERNAME: USERNAME, CONF_PASSWORD: PASSWORD},
+        )
+    assert result2["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result2["step_id"] == "user_validate"
+    assert result2["errors"] == {"base": "cannot_connect"}
+
+
+async def test_reauth_validate_invalid_login(hass: HomeAssistant):
+    """Test reauth flow with invalid username."""
+    mock_config = MockConfigEntry(
+        domain=DOMAIN,
+        unique_id=USERNAME,
+        data={
+            CONF_USERNAME: USERNAME,
+            CONF_PASSWORD: INCORRECT_PASSWORD,
+        },
+    )
+    mock_config.add_to_hass(hass)
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={
+            "source": config_entries.SOURCE_REAUTH,
+            "unique_id": mock_config.unique_id,
+        },
+        data=mock_config.data,
+    )
+    with patch(
+        "homeassistant.components.decora_wifi.common.DecoraWiFiSession",
+        side_effect=FakeDecoraWiFiSession,
+    ):
+        hass.data[DOMAIN] = {}
```

Comment:
Please don't access `hass.data` in tests

https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations